api/units.go

```diff
@@ -101,15 +101,33 @@ func (ur *unitsResource) set(rw http.ResponseWriter, req *http.Request, item str
 		return
 	}
 
+	newUnit := false
 	if eu == nil {
 		if len(su.Options) == 0 {
 			err := errors.New("unit does not exist and options field empty")
 			sendError(rw, http.StatusConflict, err)
+			return
 		} else if err := ValidateOptions(su.Options); err != nil {
 			sendError(rw, http.StatusBadRequest, err)
+			return
 		} else {
-			ur.create(rw, su.Name, &su)
+			newUnit = true
 		}
+	} else if eu.Name == su.Name && len(su.Options) > 0 {
+		// There is already a unit with the same name
+		// check the hashes if they do not match then we will
+		// create a new unit with the same name and later
+		// the job will be updated to this new unit.
```

Comment:
> > So rather than actually replacing the existing unit in the registry, you are just adding a new one with the same name? Why do you need to do it like this? By my intuition, this is bound to cause problems...
> How you think this should be done ?

My thinking was that replacing the unit in the registry should be done in one go -- ideally with an atomic operation; or if etcd doesn't offer such functionality, by deleting the old and adding the new directly after.

Note that I don't have experience with etcd, or how fleet uses it exactly. I might be missing something crucial here -- so I was hoping for some clarification in case I'm totally off...

> what problems ?

My understanding is that until now fleet wasn't ever running into situations where two units with the same name existed in the registry at the same time; and thus existing code probably isn't prepared to deal with this happening...

I might be missing something here as well of course.

> We instruct etcd to update the job key to point to the right target, and everything else is handled automatically...

Where does that happen? I don't see anything related to this in the PR; so I have to assume it happens somewhere in pre-existing code -- but where? Again, I'd appreciate some pointer, so I don't have to dig it up myself...